### PR TITLE
Migration of UserService + WorkspaceService + Tests to MySQL

### PR DESCRIPTION
**Pull request by MatthewBemis**
This is for https://broadinstitute.atlassian.net/browse/GAWB-96 and https://broadinstitute.atlassian.net/browse/GAWB-97
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [ ] **LR**: Initial review by LR and others.
- [ ] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [ ]  sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Squash commits, rebase if necessary
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Merge to develop 
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AllComponents.scala [hunk 1]
```diff
@@ -1,46 +0,0 @@
-package org.broadinstitute.dsde.rawls.dataaccess.slick
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AllComponents.scala:1]: renamed to DataAccess

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala [hunk 2]
```diff
@@ -82,6 +82,20 @@ trait RawlsBillingProjectComponent {
       }
     }
 
+    def loadAllUsersWithProjects: ReadAction[Map[RawlsUser, Iterable[RawlsBillingProjectName]]] = {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsBillingProjectComponent.scala:85]: new single query to load projects grouped by user in a single query

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponent.scala [hunk 3]
```diff
@@ -37,28 +37,28 @@ trait RawlsUserComponent {
...
-    uniqueResult[RawlsUserRecord](query).map {
-      case None => None
-      case Some(rec) => Option(unmarshalRawlsUser(rec))
+    private def loadCommon(query: RawlsUserQuery): ReadAction[Option[RawlsUser]] = {
+      uniqueResult[RawlsUserRecord](query).map {
+        case None => None
+        case Some(rec) => Option(unmarshalRawlsUser(rec))
+      }
+    }
+
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponent.scala:47]: the following were moved into the rawlsUserQuery scope

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala [hunk 4]
```diff
@@ -37,20 +39,23 @@ trait WorkflowComponent {
   import driver.api._
 
   class WorkflowTable(tag: Tag) extends Table[WorkflowRecord](tag, "WORKFLOW") {
-    def id = column[UUID]("ID", O.PrimaryKey)
+    def id = column[Long]("ID", O.PrimaryKey, O.AutoInc)
+    def externalid = column[String]("EXTERNAL_ID")
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:43]: Important change here: we no longer use the cromwell workflow id as our primary key. The initial impetus to change was because the mock remote server in our tests always serves the same workflow id. However, this is a better design because one day cromwell may decide to return us a non-unique (to us) id then boom. Or perhaps there is some use case where this is desirable. Or we need to support some execution layer other than cromwell.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala [hunk 5]
```diff
@@ -37,20 +39,23 @@ trait WorkflowComponent {
...
     def workflowEntityId = column[Option[Long]]("ENTITY_ID")
 
-    def * = (id, submissionId, status, statusLastChangedDate, workflowEntityId) <> (WorkflowRecord.tupled, WorkflowRecord.unapply)
+    def * = (id, externalid, submissionId, status, statusLastChangedDate, workflowEntityId) <> (WorkflowRecord.tupled, WorkflowRecord.unapply)
 
     def submission = foreignKey("FK_WF_SUB", submissionId, submissionQuery)(_.id)
     def workflowEntity = foreignKey("FK_WF_ENTITY", workflowEntityId, entityQuery)(_.id.?)
-  }
+
+    def uniqueWorkflowEntity = index("idx_workflow_entity", (submissionId, workflowEntityId), unique = true)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:54]: Another change: to uniquely identify a workflow within a submission the workflow entity id may be used. This is important because our primary key is not exposed in the Workflow object so this alternate key is required to identify which record a Workflow object is associated to.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala [hunk 6]
```diff
@@ -286,5 +327,10 @@ trait WorkflowComponent {
       WorkflowFailure(entity.entityName, entity.entityType, inputResolutions, errors)
     }
 
+    private def deleteMessagesAndInputs(id: Long): DBIOAction[Int, NoStream, Write with Write] = {
+//      attributeQuery.filter(_.id in (findWorkflowMessagesById(id))).delete andThen
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala]: this needs to happen but this syntax was not working

File: src/main/scala/org/broadinstitute/dsde/rawls/datamigration/AddWorkspaceGroupEmailMigration.scala [hunk 7]
```diff
@@ -1,28 +0,0 @@
-package org.broadinstitute.dsde.rawls.datamigration
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/datamigration/AddWorkspaceGroupEmailMigration.scala:1]: deleted all the migration stuff

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitor.scala [hunk 8]
```diff
@@ -49,63 +51,64 @@ object SubmissionMonitor {
...
 
   setReceiveTimeout(submissionPollInterval)
-  startWorkflowMonitorActors()
+  startWorkflowMonitorActors() pipeTo self
 
   override def receive = {
     case WorkflowStatusChange(workflow, workflowOutputs) =>
       system.log.debug("workflow state change, submission {}, workflow {} {}", submissionId, workflow.workflowId, workflow.status)
-      handleStatusChange(workflow, workflowOutputs)
+      handleStatusChange(workflow, workflowOutputs) pipeTo self
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitor.scala:67]: all these functions now return futures so pipe them to self so errors are not lost

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala [hunk 9]
```diff
@@ -1,9 +1,7 @@
 package org.broadinstitute.dsde.rawls.jobexec
 
-import cromwell.binding.types.{WdlType, WdlArrayType}
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala:3]: here begineth translation from Futures with Graphs to DBIOActions

File: src/test/scala/org/broadinstitute/dsde/rawls/integrationtest/IntegrationTestBase.scala [hunk 10]
```diff
@@ -1,137 +0,0 @@
-package org.broadinstitute.dsde.rawls.integrationtest
```
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/rawls/integrationtest/IntegrationTestBase.scala:1]: our workspace integration tests are superseded by other, better things and were too hard to fix

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponent.scala [hunk 11]
```diff
@@ -39,7 +39,7 @@ trait MethodConfigurationComponent {
     def * = (id, namespace, name, workspaceId, rootEntityType, methodNamespace, methodName, methodVersion) <> (MethodConfigurationRecord.tupled, MethodConfigurationRecord.unapply)
 
     def workspace = foreignKey("FK_MC_WORKSPACE", workspaceId, workspaceQuery)(_.id)
-    def namespaceNameIdx = index("IDX_CONFIG", (namespace, name), unique = true)
+    def namespaceNameIdx = index("IDX_CONFIG", (workspaceId, namespace, name), unique = true)
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponent.scala:42]: this had to change because it prevented the same method config from existing in multiple workspaces

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponent.scala [hunk 12]
```diff
@@ -147,6 +144,13 @@ trait MethodConfigurationComponent {
       findByWorkspace(workspaceContext.workspaceId).result.map(recs => recs.map(rec => unmarshalMethodConfigToShort(rec)))
     }
 
+    def deleteMethodConfigurationAction(id: Long) = {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponent.scala]: type signature

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala [hunk 13]
```diff
@@ -96,12 +96,23 @@ trait SubmissionComponent {
       }))
     }
 
+    def listWithUser(workspaceContext: SlickWorkspaceContext): ReadAction[Seq[(Submission, RawlsUser)]] = {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala]: I'd prefer listWithSubmitter

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala [hunk 14]
```diff
@@ -164,6 +175,13 @@ trait SubmissionComponent {
       }))
     }
 
+    def deleteSubmissionAction(submissionId: UUID) = {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala]: return type signature

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala (hunk 6 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala]: Do you mean this is a TODO?

File: src/main/scala/org/broadinstitute/dsde/rawls/monitor/BucketDeletionMonitor.scala [hunk 15]
```diff
@@ -1,14 +1,24 @@
...
 import org.broadinstitute.dsde.rawls.model.PendingBucketDeletions
 import org.broadinstitute.dsde.rawls.monitor.BucketDeletionMonitor.{BucketDeleted, DeleteBucket}
+import org.broadinstitute.dsde.rawls.dataaccess.SlickDataSource
+import org.broadinstitute.dsde.rawls.dataaccess.slick.DataAccess
+import org.broadinstitute.dsde.rawls.dataaccess.slick.ReadAction
+import org.broadinstitute.dsde.rawls.dataaccess.slick.PendingBucketDeletionRecord
+import org.broadinstitute.dsde.rawls.dataaccess.slick.PendingBucketDeletionRecord
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import org.broadinstitute.dsde.rawls.dataaccess.slick.PendingBucketDeletionRecord
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/monitor/BucketDeletionMonitor.scala]: triple import

File: src/main/scala/org/broadinstitute/dsde/rawls/monitor/BucketDeletionMonitor.scala [hunk 16]
```diff
@@ -16,30 +26,27 @@ object BucketDeletionMonitor {
...
+    case BucketDeleted(bucketName) => recordBucketDeletion(bucketName) pipeTo self
+    
+    case Unit => // successful future
+    case Status.Failure(t) => logger.error("failure in BucketDeletionMonitor", t)
   }
 
-  private def getPendingDeletions(txn: RawlsTransaction) =
-    containerDAO.workspaceDAO.loadPendingBucketDeletions(txn) getOrElse {
-      throw new RawlsException("Cannot find the list of buckets pending deletion")
+  private def deleteBucket(bucketName: String): Future[Unit] = {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/monitor/BucketDeletionMonitor.scala:37]: These two have accidentally gotten reversed.  The logic should go:

Delete Bucket = SAVE to PBD table, tell Google to delete

Record Deletion = REMOVE from PDB table

File: src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala [hunk 17]
```diff
@@ -513,6 +533,7 @@ class SubmissionSpec(_system: ActorSystem) extends TestKit(_system) with FlatSpe
   }
 
   "Getting workflow metadata" should "return 200" in withSubmissionTestWorkspaceService { workspaceService =>
+    println(subTestData.submissionTestAbortGoodWorkflow)
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala]: stray println

File: src/test/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitorSpec.scala [hunk 18]
```diff
@@ -83,26 +81,9 @@ class WorkflowMonitorSpec(_system: ActorSystem) extends TestKit(_system) with Fl
     }
   }
 
-  it should "fail a workflow if the method config can't be found" in withDefaultTestDatabase { dataSource =>
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitorSpec.scala:86]: disable and comment instead of delete

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/AdminApiServiceSpec.scala [hunk 19]
```diff
@@ -33,22 +33,14 @@ class AdminApiServiceSpec extends ApiServiceSpec {
...
-  def getMatchingBillingProjectVertices(dataSource: DataSource, project: RawlsBillingProject): Iterable[Vertex] =
-    dataSource.inTransaction() { txn =>
-      txn.withGraph { graph =>
-        graph.getVertices.filter(v => {
-          v.asInstanceOf[OrientVertex].getRecord.getClassName.equalsIgnoreCase(VertexSchema.BillingProject) &&
-            v.getProperty[String]("projectName") == project.projectName.value
-        })
-      }
-    }
+  def getMatchingBillingProjectVertices(dataSource: SlickDataSource, project: RawlsBillingProject) = runAndWait(rawlsBillingProjectQuery.load(project.projectName))
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/AdminApiServiceSpec.scala]: rename: not vertices any more

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiServiceSpec.scala [hunk 20]
```diff
@@ -409,7 +377,7 @@ class EntityApiServiceSpec extends ApiServiceSpec {
       }
   }
 
-  val attributeList = AttributeValueList(Seq(AttributeString("a"), AttributeString("b"), AttributeBoolean(true)))
+  val attributeList = AttributeValueList(Seq(AttributeString("a"), AttributeString("b"), AttributeString("c")))
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiServiceSpec.scala:380]: why was this changed?  We should be sure mixed-type lists like this work, if we support them

File: src/test/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitorSpec.scala (hunk 18 above)
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitorSpec.scala:86]: I removed it because the behavior has changed to the point where the test does not make sense anymore.

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiServiceSpec.scala (hunk 20 above)
Comment (MatthewBemis) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiServiceSpec.scala:380]: we do not support mixed-type lists anymore

File: src/main/scala/org/broadinstitute/dsde/rawls/monitor/BucketDeletionMonitor.scala (hunk 15 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/monitor/BucketDeletionMonitor.scala]: haha

File: src/test/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitorSpec.scala (hunk 18 above)
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitorSpec.scala:86]: ok

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiServiceSpec.scala (hunk 20 above)
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiServiceSpec.scala:380]: ok

**Comment by jmthibault79** [comment]
Done with review except for WorkspaceService.  Attempting review of that via local IntelliJ diff.

**Comment by jmthibault79** [comment]
WorkspaceService comments: 
- triple import of PerRequestMessage
- updateWorkspace / updateEntity : rewrite with Try
- loadFromEmail: duplicates RawlsGroupComponent's function.  remove one
- line 427: comment extensively on this foldLeft
- lock/unlock Workspace: make new workspaceQuery lock/unlock methods instead of resaving workspace
- line 1054-1056: this is the toFutureTry pattern.  Can replace with 

```
            val outputFTs = toFutureTry(executionServiceDAO.outputs(workflowId, userInfo))
            val logFTs = toFutureTry(executionServiceDAO.logs(workflowId, userInfo))
            DBIO.from(outputFTs zip logFTs map { 
```

and be sure to remove the matching `}`
- hasAllUserReadAccess: match on readerGroup instead of using get
- getWorkspaceStatus: remove run/Await pattern
- can remove 1377 "note: executes in a Future"

**Comment by MatthewBemis** [comment]
All non-WorkspaceService related comments have been addressed

**Comment by MatthewBemis** [comment]
PR comments addressed except for:

https://github.com/broadinstitute/rawls/pull/341#discussion_r56109108
and the following in WorkspaceService:
- updateWorkspace / updateEntity : rewrite with Try
- line 427: comment extensively on this foldLeft
- getWorkspaceStatus: remove run/Await pattern

**Comment by dvoet** [comment]
@jmthibault79 review comments addressed

**Comment by jmthibault79** [comment]
@dvoet could you respond to my comment on Workflow Component line 331?

**Comment by jmthibault79** [comment]
hasAllUserReadAccess error text is incorrect: the Workspace's reader group is not found

**Comment by jmthibault79** [comment]
Looks good except for these 2 comments

**Comment by MatthewBemis** [comment]
Crap, that last one is my bad :sob:
